Implement NodeAddressing on top of Table[NodeAddress]

joamaki opened the pull request:
This works towards making the cilium-agent responsive to node address changes by moving from global variables and direct use of `netlink.AddrList` to using statedb tables that can be queried and watched for changes. 

Reimplement `NodeAddressing` to query against `Table[NodeAddress]` (derived from `Table[*Device]` which is populated from netlink subscriptions to links, addresses and routes) and `LocalNodeStore` instead of using the pkg/node globals or doing a `netlink.AddrList`. Eventual goal is to remove `NodeAddressing` and instead have the modules query `Table[NodeAddress]` directly and reconcile if it changes.

Remove the `ipv{4,6}NodePortAddrs` globals and the related getters and init, replacing uses of them with either `NodeAddressing` or queries to `Table[NodeAddress]`.

See commit messages for more details.

joamaki committed:
datapath: Reimplement NodeAddressing around Table[NodeAddress]

As a step towards removing the global address maps in pkg/node
(ipv4NodeportAddrs, ipv4MasqAddrs etc.), reimplement NodeAddressing
as queries against the Table[NodeAddress].

The goal is to eventually remove NodeAddressing and have modules
query the Table[NodeAddress] directly and to react to changes, but
as an intermediary step without having to refactor whole bunch of
uses of NodeAddressing, this still serves the purpose of allowing
removal of some global variables and removing the other ways of
extracting the device and address information from the system.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

joamaki committed:
agent: Remove ipv4NodePortAddrs and friends

Remove the ipv{4,6}NodePortAddrs global variables and associated
getters and inits from pkg/node and replace their uses with Table[NodeAddress]
and NodeAddressing (which queries from the former).

This paves the path towards being able to react to the local node's address
changes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

joamaki committed:
datapath: Simplify NodeAddressing and add DirectRouting method

Use a single set of methods for both IPv4 and IPv6 to remove
code duplication.

Add DirectRouting() method for looking up the direct routing device
and IP address.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

joamaki commented on pkg/datapath/tables/node_addressing.go:
```diff
@@ -0,0 +1,141 @@
+	return a.getNodeAddresses(a.db.ReadTxn(), a.flags), nil
+}
+
+func (a addressFamily) DirectRouting() (int, net.IP, bool) {
```
This is likely half-implemented with not all uses of `option.Config.DirectRoutingDevice` converted.

thorn3r commented on pkg/datapath/tables/node_addressing.go:
```diff
@@ -0,0 +1,145 @@
+	deviceName := option.Config.DirectRoutingDevice
+	if deviceName == "" {
+		return 0, nil, false
+	}
```
nit: unused local
```suggestion
	if option.Config.DirectRoutingDevice == "" {
		return 0, nil, false
	}
```

thorn3r reviewed (commented):
small nit on a 1st pass, I see we're working out some failures/merge conflicts so I'll take another look later

bimmlerd committed:
node_address: workaround w/ overlapping link names

Before this patch, the creation of a link named 'cilium' without
addresses would delete node addresses added by a link called
'cilium_host'.

The issue stems from the fact that statedb Get queries on variable
length keys perform prefix searches, not exact queries. As a result, the
'existing' IPs that the node address table returned when queried for
device name 'cilium' included addresses for 'cilium_host', which were
not in 'new' (the set of addresses on 'cilium'). The update function
then threw out these existing addresses, as it considers these removed
from the device.

Arguably, this is a statedb bug (c.f cilium/cilium#29324), but we can
work around it here.

Fixes: 5342d0104f (datapath/tables: Add Table[NodeAddress])

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
node_address: handle link scoped cilium_host addrs

In the transition to the table, a bit of logic handling the specific
case of link scoped host device addresses got lost. Reinstate it, as
without it, the certain cilium_host addresses which were previously
considered node addresses were no longer considered node addresses.

The impact of this was that FIB lookups into the endpoint maps would
fail for pod to host traffic in the bpf host routing path.

Fixes: 5342d0104f (datapath/tables: Add Table[NodeAddress])

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

ti-mo commented on pkg/datapath/loader/loader.go:
```diff
@@ -325,7 +325,7 @@ func (l *Loader) reloadHostDatapath(ctx context.Context, ep datapath.Endpoint, o
 	if _, err := netlink.LinkByName(defaults.SecondHostDevice); err != nil {
 		log.WithError(err).WithField("device", defaults.SecondHostDevice).Error("Link does not exist")
-		return err
+		return fmt.Errorf("device '%s' not found: %w", defaults.SecondHostDevice, err)
```
Nit: this is really unlikely to happen, as we create this device a little bit before this, but I guess that's where it went wrong while coming up with this patch. :)

ti-mo commented on pkg/datapath/loader/hash_test.go:
```diff
@@ -20,7 +27,21 @@ var (
+	hv := hive.New(
+		cell.Provide(
+			fake.NewNodeAddressing,
+			func() bandwidth.Manager { return &fake.BandwidthManager{} },
```
Why is the bandwidth manager a requirement here?

thorn3r commented on pkg/datapath/tables/node_address.go:
```diff
@@ -291,14 +302,14 @@ func (n *nodeAddressController) getAddressesFromDevice(dev *Device) (addrs sets.
 		if n.AddressScopeMax < unix.RT_SCOPE_LINK {
 			for _, addr := range sortedAddresses(dev.Addrs) {
 				if addr.Scope == unix.RT_SCOPE_LINK {
+					n.Log.Infof("HII2 - adding addr: %v prim: %v np: %v dev: %v", addr.Addr, !addr.Secondary, false, dev.Name)
```
is this leftover from debugging? If not, maybe something more descriptive (HII2 doesn't seem very descriptive without looking at the code)

thorn3r commented on pkg/datapath/tables/node_address.go:
```diff
@@ -302,7 +302,6 @@ func (n *nodeAddressController) getAddressesFromDevice(dev *Device) (addrs sets.
 			for _, addr := range sortedAddresses(dev.Addrs) {
 				if addr.Scope == unix.RT_SCOPE_LINK {
-					n.Log.Infof("HII2 - adding addr: %v prim: %v np: %v dev: %v", addr.Addr, !addr.Secondary, false, dev.Name)
 					addrs.Insert(NodeAddress{
```
ah, I see they are removed here in a later commit

thorn3r reviewed (approved):
Nice work :+1:

aditighag reviewed (approved):
Approving for pkg/endpoint changes.

ysksuzuki commented on daemon/cmd/daemon.go:
```diff
@@ -1129,13 +1135,9 @@ func (d *Daemon) ReloadOnDeviceChange(devices []string) {
 	}
 
 	if option.Config.EnableNodePort {
-		if err := node.InitNodePortAddrs(devices, option.Config.LBDevInheritIPAddr); err != nil {
```
`LBDevInheritIPAddr` is no longer referenced. Does this change ignore  `bpf-lb-dev-ip-addr-inherit` flag users specified?

joamaki replied:
Will fix this in follow-up PR once we've figured out the right way to do it.
